astropy/units/_typing.py

```diff
@@ -0,0 +1,195 @@
+# -*- coding: utf-8 -*-
+
+"""Typing compatibility for :mod:`~astropy.units`.
+
+Static types are imported from
+
+1. ``typing`` (if present)
+2. ``typing_extensions`` (if present)
+3. Here if not present in the above.
+   These are altered from https://github.com/python/typing
+
+"""
+
+__all__ = [
+    "Annotated",
+]
+
+import typing
+import operator
+
+try:  # py 3.9+
+    from typing import Annotated
+except ImportError:  # optional dependency
+    try:
+        from typing_extensions import Annotated
+    except ImportError:
+        _HAS_ANNOTATED = False   # need to do ourselves
+    else:
+        _HAS_ANNOTATED = True
+else:
+    _HAS_ANNOTATED = True
+
+
+NoneType = type(None)
+
+
+def _isAnnotated(t) -> bool:
+    """Is this thing an annotated type?
+
+    Checks that the object has attributes origin and metadata
+    and that it passes an equality test with Annotated[origin, metadata].
+    This test is designed to pass if ``Annotated`` is from
+    :mod:`~typing`, ``typing_extensions``, or our compatibility class.
+
+    .. todo::
+
+        When min python version is 3.9+ or typing_extensions is a run-time
+        dependency, I think there's a built-in method to test if something
+        is an Annotated.
+
+    Parameters
+    ----------
+    t : object
+
+    Returns
+    -------
+    bool
+        Whether the object is an ``Annotated``.
+
+    """
+    if not hasattr(t, "__origin__") or not hasattr(t, "__metadata__"):
+        return False
+
+    # we need to construct the "expected" Annotation with the underlying
+    # class getitem to avoid having to special-case single annotations.
+    # Annotated turns annotations into a tuple, so we want to ensure
+    # correct unpacking, which can only be done with the function syntax,
+    # not the getitem [].
+    expected = Annotated.__class_getitem__((t.__origin__, *t.__metadata__))
+
+    if t == expected:
+        return True
+
+    return False
+
+
+def _isUnion(t) -> bool:
+    """Check whether a type is a Union."""
+    if hasattr(t, "__origin__"):
```

Comment:
I'd write this myself as `return getattr(t, '__origin__', None) is typing.Union` - which suggests it is perhaps not quite worth defining it for a single use in the code...

Reply:
Now that I've moved `_typing` to extern, I think this might be useful in astropy.modeling. But I'll remove this until then.

Reply:
✅